ext/pdo/pdo_stmt.c

```diff
@@ -772,88 +789,78 @@ static bool do_fetch(pdo_stmt_t *stmt, zval *return_value, enum pdo_fetch_type h
-				old_arg_count = stmt->fetch.cls.fci.param_count;
-				do_fetch_opt_finish(stmt, 0);
-
-				fetch_value(stmt, &val, column_index_to_fetch++, NULL);
-				if (Z_TYPE(val) != IS_NULL) {
-					if (!try_convert_to_string(&val)) {
-						return 0;
-					}
-					if ((cep = zend_lookup_class(Z_STR(val))) == NULL) {
-						stmt->fetch.cls.ce = ZEND_STANDARD_CLASS_DEF_PTR;
-					} else {
-						stmt->fetch.cls.ce = cep;
+				zval ce_name_from_column;
+				fetch_value(stmt, &ce_name_from_column, column_index_to_fetch++, NULL);
+				if (Z_TYPE(ce_name_from_column) == IS_STRING) {
+					ce = zend_lookup_class(Z_STR(ce_name_from_column));
+					if (ce == NULL) {
+						ce = zend_standard_class_def;
 					}
 				}
-
-				do_fetch_class_prepare(stmt);
-				zval_ptr_dtor_str(&val);
-			}
-			ce = stmt->fetch.cls.ce;
-			/* TODO: Make this an assertion and ensure this is true higher up? */
-			if (!ce) {
-				/* TODO Error? */
-				pdo_raise_impl_error(stmt->dbh, stmt, "HY000", "No fetch class specified");
-				return 0;
+				/* Use default CE if present */
+				if (ce == NULL) {
+					ce = zend_standard_class_def;
+				}
+				zval_ptr_dtor_str(&ce_name_from_column);
+			} else {
+				/* This can happen if the fetch flags are set via PDO::setAttribute()
+				 * $pdo->setAttribute(PDO::ATTR_DEFAULT_FETCH_MODE, PDO::FETCH_CLASS);
+				 * See ext/pdo/tests/bug_38253.phpt */
+				if (UNEXPECTED(ce == NULL)) {
+					pdo_raise_impl_error(stmt->dbh, stmt, "HY000", "No fetch class specified");
+					return false;
+				}
+				ctor_arguments = stmt->fetch.cls.ctor_args;
 			}
-			if ((flags & PDO_FETCH_SERIALIZE) == 0) {
-				if (UNEXPECTED(object_init_ex(return_value, ce) != SUCCESS)) {
-					return 0;
+			ZEND_ASSERT(ce != NULL);
+			if (flags & PDO_FETCH_SERIALIZE) {
+				if (!ce->unserialize) {
+					// TODO, should we mention the class? Considering this option is broken and deprecated...
+					pdo_raise_impl_error(stmt->dbh, stmt, "HY000", "cannot unserialize class");
+					return false;
 				}
-				if (!stmt->fetch.cls.fci.size) {
-					if (!do_fetch_class_prepare(stmt)) {
-						zval_ptr_dtor(return_value);
-						return 0;
-					}
+			} else {
+				if (UNEXPECTED(object_init_ex(return_value, ce) != SUCCESS)) {
+					return false;
 				}
-				if (ce->constructor && (flags & PDO_FETCH_PROPS_LATE)) {
-					zval retval_constructor_call;
-					stmt->fetch.cls.fci.retval = &retval_constructor_call;
-					stmt->fetch.cls.fci.object = Z_OBJ_P(return_value);
-					stmt->fetch.cls.fcc.object = Z_OBJ_P(return_value);
-					zend_call_function(&stmt->fetch.cls.fci, &stmt->fetch.cls.fcc);
-					if (Z_TYPE(retval_constructor_call) == IS_UNDEF) {
-						/* Exception has happened */
+				if (ce->constructor && flags & PDO_FETCH_PROPS_LATE) {
+					bool failed = pdo_call_fetch_object_constructor(ce->constructor, ctor_arguments, return_value);
+					if (UNEXPECTED(failed)) {
 						zval_ptr_dtor(return_value);
+						RETVAL_FALSE;
```

Comment:
I don't think this line is necessary? We're already signaling failure and the callers seem to handle this. The choice of RETVAL_FALSE instead of something else like NULL/UNDEF also strikes me as weird in this regard.

Reply:
Okay let's just make it consistently not use it then as it's a big refactor anyway it seems like a good opportunity to do so.